### ISSUE TITLE
Add support for bindAddress in the TCP4ClientEndpoint description parser

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6465]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)| @wallrj reported|
|-|-|
|Trac ID|trac#6465|
|Type|enhancement|
|Created|2013-04-23 18:21:24Z|
|Branch|https://github.com/twisted/twisted/tree/tcpclient-endpoint-bindaddress-6465|

clientFromString accepts a bindAddress keyword but it fails to
properly parse the value.

The string value instead of a tuple(ip, port) is passed to connectTCP
resulting in an Exception.

Here's a demo.

```python
from twisted.internet.endpoints import clientFromString
from twisted.internet.task import react
from twisted.internet.protocol import ClientFactory
from twisted.protocols.basic import LineReceiver



def main(reactor):
    f = ClientFactory()
    f.protocol = LineReceiver
    e = clientFromString(reactor, 'tcp:127.0.0.1:8080:bindAddress=127.0.0.0')
    return e.connect(f)



if __name__ == '__main__':
    react(main, ())

```

```
$ python clientbindaddress.py
main function encountered error
Traceback (most recent call last):
  File "clientbindaddress.py", line 17, in <module>
    react(main, ())
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/task.py", line 830, in react
    finished = main(_reactor, *argv)
  File "clientbindaddress.py", line 12, in main
    return e.connect(f)
--- <exception caught here> ---
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/endpoints.py", line 543, in connect
    timeout=self._timeout, bindAddress=self._bindAddress)
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/posixbase.py", line 496, in connectTCP
    c.connect()
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/base.py", line 1046, in connect
    self.transport = transport = self._makeTransport()
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/tcp.py", line 1174, in _makeTransport
    return Client(self.host, self.port, self.bindAddress, self, self.reactor)
  File "/home/richard/projects/Twisted/branches/idatagram-endpoint-4471/twisted/internet/tcp.py", line 702, in __init__
    skt.bind(bindinfo)
  File "/usr/lib64/python2.7/socket.py", line 224, in meth
    return getattr(self._sock,name)(*args)
exceptions.TypeError: getsockaddrarg: AF_INET address must be tuple, not str

```

I think you should be able to omit the bindAddress port from the
string and the parser should use 0 by default.

I guess the bind IP address must always be the same family (IPv4,IPv6)
as the target address.

And perhaps bindAddress should also be considered in [#4859](https://github.com/twisted/twisted/issues/4859). eg If you
choose an IPv4 bindAddress there's no point looking up AAAA records
etc.

Not sure if a separate ticket needs to be raised for
TCP6ClientEndpoint description parser.

<details><summary>Searchable metadata</summary>

```
trac-id__6465 6465
type__enhancement enhancement
reporter__rwall rwall
priority__normal normal
milestone__ 
branch__branches_tcpclient_endpoint_bindaddress_6465 branches/tcpclient-endpoint-bindaddress-6465
branch_author__rwall rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1366741284000000 1366741284000000
changetime__1370881488000000 1370881488000000
version__None None
owner__rwall rwall

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

> I think you should be able to omit the bindAddress port from the string and the parser should use 0 by default.

Good idea.  That sounds handy and convenient.  How will you specify the port if you want that behavior?  Probably not "127.0.0.1:123".  It would need to be "127.0.0.1\:123" at least, which is kinda sad.  Maybe another keyword argument?

And maybe being able to specify the port is a separate feature from this ticket anyway?

> And perhaps bindAddress should also be considered in [[#4859](https://github.com/twisted/twisted/issues/4859)](https://github.com/twisted/twisted/issues/4859). eg If you choose an IPv4 bindAddress there's no point looking up AAAA records etc.

That might be an interesting optimization, but doesn't need to be part of #4859.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj commented|
|-|-|

(In [38314]) Branching to 'tcpclient-endpoint-bindaddress-6465'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_4"></a>|@wallrj removed owner<br>@wallrj set status to `new`|
|-|-|

Ready for review in log:branches/tcpclient-endpoint-bindaddress-6465
 1. Modified _parseClientTCP to check coerce the bindAddress into an
    address tuple with port fixed to 0.
 2. There were existing tests to exercise bindAddress so I modified
    them to expect the tuple result.
    1. The existing test methods contain many assertions. I was
       tempted to remove the bindAddress from the endpoint
       descriptions in existing tests and create separate bindAddress
       related test methods. Good idea or not?
 3. I added a note about bindAddress to the API docs because I guess
    people who are used to the IReactor.connectTCP method may expect
    to be able to pass (address,port) in the description string.
    1. In fact for that reason, I wonder if it would be better to use
       a different keyword eg bindAddressIP?
    2. Or maybe it would be even clearer if we called it 'interface'
       which would give nice symmetry with the serverFromString
       syntax?
 4. As exarkun suggested (above) we can add another keyword later if
    anyone really needs to specify the client port. eg bindAddressPort
    1. Alternatively we could use a different delimiter eg
       bindAddress=127.0.0.1@1234 or 127.0.0.1,1234.
 5. Some random thoughts
    1. I know there isn't a tcp6 endpoint description parser yet, but
       is it going to be possible to parse the ":"s in IPv6 addresses
       without lots of horrible escaping?
    2. It would be nice if the clientFromString and serverFromString
       methods raised dedicated parsing exceptions with nice clear
       error messages which could be eg printed to stderr in the case
       of bad endpoint descriptions provided in command line
       arguments.

Build results look ok:
 * http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/tcpclient-endpoint-bindaddress-6465

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun commented|
|-|-|

> I know there isn't a tcp6 endpoint description parser yet, but is it going to be possible to parse the ":"s in IPv6 addresses without lots of horrible escaping?

There are TCP6 endpoint description parsers.  And they work how you fear, eg "fe80\:\:224\:d6ff\:fe53\:ede".  This is why I cautioned against introducing a new, unconsidered syntax in [#6451](https://github.com/twisted/twisted/issues/6451).

> It would be nice if the clientFromString and serverFromString methods raised dedicated parsing exceptions with nice clear error messages which could be eg printed to stderr in the case of bad endpoint descriptions provided in command line arguments.

_Something_ should do this, but it doesn't have to be `clientFromString` and `serverFromString`.  The fiddly parsing implementation does not have to be mixed together with user interface considerations - and _should_ not be in most cases, in order to keep the code simple and maintainable and avoid forcing the user interface down a certain predefined path (eg, maybe your user interface is a GUI: you don't want an english language description of the problem, you want character offsets so you can highlight the broken part of the input).

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @wallrj|
|-|-|

OK, sounds reasonable to me. Please add a NEWS fragment, open a bug to add bindAddressPort, and merge. Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_7"></a>|@wallrj set status to `closed`|
|-|-|

(In [38706]) Merge tcpclient-endpoint-bindaddress-6465: Parse the bindAddress keyword in TCP related client endpoint strings

Author: rwall
Reviewer: therve,exarkun
Fixes: [#6465](https://github.com/twisted/twisted/issues/6465)
Refs: [#6562](https://github.com/twisted/twisted/issues/6562)

Parse the bindAddress keyword argument in TCP related client endpoint
strings. The value is treated as a host string and is coerced to a
(host, port) tuple. The port is set to 0 to bind the client to any
ephemeral port.
